### PR TITLE
WT-13013 Add API call count instrumentation to WiredTiger

### PR DESCRIPTION
**Pull request by luke-pearson**
<### Remove after completion

###>

When configured WiredTiger will dump data like this:
```
$ cat debug_api_call_counts.txt
WT_CURSOR::get_key=2699053
WT_CURSOR::get_value=4155908
WT_CURSOR::get_raw_key_value=2697982
WT_CURSOR::set_key=9445989
WT_CURSOR::set_value=232
WT_CURSOR::compare=0
WT_CURSOR::equals=0
WT_CURSOR::next=2810406
WT_CURSOR::prev=358
WT_CURSOR::reset=14987970
WT_CURSOR::search=2698193
WT_CURSOR::search_near=786
WT_CURSOR::insert=228
WT_CURSOR::insert_check=24
WT_CURSOR::modify=1351200
WT_CURSOR::update=0
WT_CURSOR::remove=0
WT_CURSOR::reserve=0
WT_CURSOR::checkpoint_id=0
WT_CURSOR::close=5617560
WT_CURSOR::largest_key=3
WT_CURSOR::reconfigure=0
WT_CURSOR::bound=2697989
WT_CURSOR::cache=0
WT_CURSOR::reopen=0
WT_SESSION::close=94
WT_SESSION::reconfigure=81
WT_SESSION::strerror=0
WT_SESSION::open_cursor=5391197
WT_SESSION::alter=0
WT_SESSION::bind_configuration=0
WT_SESSION::create=1
WT_SESSION::compact=0
WT_SESSION::drop=0
WT_SESSION::join=0
WT_SESSION::log_flush=921
WT_SESSION::log_printf=0
WT_SESSION::reset=2692670
WT_SESSION::salvage=0
WT_SESSION::truncate=0
WT_SESSION::upgrade=0
WT_SESSION::verify=0
WT_SESSION::begin_transaction=2697996
WT_SESSION::commit_transaction=1345784
WT_SESSION::prepare_transaction=0
WT_SESSION::rollback_transaction=1352212
WT_SESSION::query_timestamp=0
WT_SESSION::timestamp_transaction=0
WT_SESSION::timestamp_transaction_uint=0
WT_SESSION::checkpoint=2
WT_SESSION::reset_snapshot=0
WT_SESSION::transaction_pinned_range=2692670
WT_SESSION::get_rollback_reason=0
WT_SESSION::breakpoint=0
WT_CONNECTION::close=1
WT_CONNECTION::debug_info=0
WT_CONNECTION::reconfigure=0
WT_CONNECTION::get_home=0
WT_CONNECTION::compile_configuration=31
WT_CONNECTION::configure_method=0
WT_CONNECTION::is_new=0
WT_CONNECTION::open_session=79
WT_CONNECTION::query_timestamp=94
WT_CONNECTION::set_timestamp=0
WT_CONNECTION::rollback_to_stable=0
WT_CONNECTION::load_extension=0
WT_CONNECTION::add_data_source=0
WT_CONNECTION::add_collator=0
WT_CONNECTION::add_compressor=4
WT_CONNECTION::add_encryptor=0
WT_CONNECTION::add_extractor=0
WT_CONNECTION::set_file_system=0
WT_CONNECTION::add_storage_source=0
WT_CONNECTION::get_storage_source=0
WT_CONNECTION::get_extension_api=0
```

### CODE REVIEW

File: dist/api_call_count.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                if re.search(r"^func", func.group(1)):
+                    continue
+                # We've found an API function.
+                api_func_h.write("#define WT_API_" + api_string + "_" + func.group(1) + " " + str(func_count) + "\n")
+                append_api_name(get_name(api_string, func.group(1)), first)
+                if first:
+                    first = False
+                func_count += 1
+                # Handle some fake API wiredtiger has.
+                if api_string == "WT_CURSOR" and func.group(1) == "insert":
```
Comment (luke-pearson) [inline dist/api_call_count.py:58]: Here

File: dist/api_call_count.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+first = True
+with open('../src/include/api_func_map.h', "w") as api_func_h:
+    api_func_h.write("/* This is an auto generated file, do not touch! */\n")
+    api_func_h.write("#pragma once\n")
+    with open('../src/include/wiredtiger.in') as api_def:
+        for line in api_def:
+            struct = re.search(r"struct ([a-z_]+) \{", line)
+            if struct is not None:
+                if re.search(r"__wt_connection", struct.group(1)):
```
Comment (tod-johnson-mongodb) [inline dist/api_call_count.py:38]: String equality == would work and be faster. The same for new lines 40 and 42.

File: dist/api_call_count.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python3
+import os, re, sys, textwrap
+
+def get_name(type, func):
+    return "\"" + type + "::"   + func + "\""
+
+api_name_lookup = ["const char *__WT_API_NAMES[] = {"]
+api_name_line_count = 0
+
+# Cursed code that outputs the string array in the correctly formatted fashion.
```
Comment (tod-johnson-mongodb) [inline dist/api_call_count.py:10]: Change "Cursed code that outputs" to "Output"

File: dist/api_call_count.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                # Handle some fake API wiredtiger has.
+                if api_string == "WT_CURSOR" and func.group(1) == "insert":
+                    api_func_h.write("#define WT_API_WT_CURSOR_insert_check " + str(func_count) + "\n")
+                    append_api_name(get_name('WT_CURSOR', 'insert_check'), False)
+                    func_count += 1
+    api_func_h.write("#define WT_API_COUNT " + str(func_count) + "\n")
+    api_func_h.write('extern const char* __WT_API_NAMES[];\n')
+
+api_name_lookup[api_name_line_count] += "};\n"
+with open('../src/config/api_func_map.c', 'w') as api_func_c:
```
Comment (tod-johnson-mongodb) [inline dist/api_call_count.py:66]: 1. Where is the code that starts the file with #include "api_func_map.h"?
2. api_func_map.c should have a comment saying it is auto generated.

File: src/conn/conn_api.c [hunk 5]
```diff
@@ -2128,6 +2157,31 @@ __debug_mode_log_retention_config(WT_SESSION_IMPL *session, const char *cfg[])
...
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+
+    WT_RET(__wt_config_gets(session, cfg, "debug_mode.api_call_count", &cval));
+    if (cval.val) {
+        FLD_SET(conn->debug_flags, WT_CONN_DEBUG_API_ENTRY_COUNT);
+        /* Allocate the API call count tracker. */
+        WT_RET(__wt_calloc_one(session, &conn->api_call_tracker));
+    } else if (!FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_API_ENTRY_COUNT)) {
```
Comment (tod-johnson-mongodb) [inline src/conn/conn_api.c:2177]: Change !FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_API_ENTRY_COUNT)
to FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_API_ENTRY_COUNT)

**Review by tod-johnson-mongodb — APPROVED** [review]
There are a few changes that are needed: conn_api.c new line 2177, and api_call_count.py new line 66. The rest are suggestions.

I am approving this since I won't be here to respond next week.

**Comment by etienneptl** [comment]
Hi @luke-pearson, is this PR still going?

**Pull request closed without merging**